### PR TITLE
Delete user related oauth stuff on user deletion too

### LINKED ISSUE #19111: Deleted accounts leave OAuth tokens valid
### Gitea Version

Gitea version 1.16.3 built with GNU Make 4.1, go1.17.7 : bindata, sqlite, sqlite_unlock_notify

### Git Version

2.25.1

### Operating System

Ubuntu 20.04

### How are you running Gitea?

Installed with:

```
# mkdir 
# useradd -m -r -d /srv/gitea gitea
# curl -L https://github.com/go-gitea/gitea/releases/download/v1.16.3/gitea-1.16.3-linux-amd64 -o /srv/gitea/gitea
# chmod +x /srv/gitea/gitea
```

Run manually with:

```
# sudo -i gitea
# ./gitea
```

On:

```
# cat /etc/os-release 
NAME="Ubuntu"
VERSION="20.04.4 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.4 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
```

(and I have an nginx reverse proxy for TLS in front; if you want to see it's config too I can add it)

### Database

PostgreSQL

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

https://gist.github.com/kousu/7a0adf177e50bf8387a00b4fa3fd7dd3

### Description

An OAuth app created by a deleted user is left behind in the database, and tokens it handed out are still valid, and can even see **private repos** created **after** the deletion. 

### Reproduction

1. Make two accounts

    <details><summary>Account 1</summary>

    Screenshot 2022-03-16 at 19-03-49 Gitea Git with a cup of tea

    </details>

    <details><summary>Account 2</summary>

    Screenshot 2022-03-16 at 19-04-00 Gitea Git with a cup of tea

    </details>

2. In account 1: create an OAuth app

    <details>

    Screenshot 2022-03-16 at 19-20-09 Gitea Git with a cup of tea

    </details>

3. In account 2: create a couple repositories

    <details>

    Screenshot 2022-03-16 at 20-43-22 bivott82

    </details>

    _You can see this on the demo site at_: https://try.gitea.io/bivott82

2. Deploy DroneCI (or some other OAuth consumer), providing it the OAuth app credentials.

    Get a public IP, set up DNS and open firewall ports for :80 and :443, and install docker before running this:

    ```
    docker run \
      --volume=/var/lib/drone:/data \
      --env=DRONE_GITEA_SERVER=https://try.gitea.io \
      --env=DRONE_GITEA_CLIENT_ID=4d9b78ea-d8fc-45ed-a11a-ea684a690108 \
      --env=DRONE_GITEA_CLIENT_SECRET=gto_z5tbfeunl4g752hlj5aeoksj25jev7lxblzejdop5sxt76nmacvq \
      --env=DRONE_RPC_SECRET="$(openssl rand -hex 16)" \
      --env=DRONE_SERVER_HOST=drone1.kousu.ca \
      --env=DRONE_SERVER_PROTO=https \
      --env=DRONE_TLS_AUTOCERT=true \
      --publish=80:80 \
      --publish=443:443 \
      --restart=always \
      --detach=true \
      --name=drone \
      drone/drone:2
    ```

5. In account 2: login to DroneCI; you will see the two repositories:

    <details>

    Screenshot 2022-03-16 at 20-44-09 Drone CI

    Screenshot 2022-03-16 at 20-44-19 Gitea Git with a cup of tea

    Screenshot 2022-03-16 at 20-44-46 Drone CI
    Screenshot 2022-03-16 at 20-44-53 Drone CI

    </details>

3. In account 1: delete account

    <details>

    Screenshot 2022-03-16 at 20-48-38 Gitea Git with a cup of tea

    </details>

5. In account 1: create a third repository

    <details>

    Screenshot 2022-03-16 at 20-49-08 Gitea Git with a cup of tea

    </details>

    _You can see this on the demo site at_: https://try.gitea.io/bivott82

7. In DroneCI: refresh

    <details>

    Screenshot 2022-03-16 at 20-49-33 Drone CI

    </details>

   You will: **incorrectly** see the third repository.

    <details>

    Screenshot 2022-03-16 at 20-49-44 Drone CI

    </details>

9. Log out of DroneCI

    <details>

    Screenshot 2022-03-16 at 20-49-59 Account - Drone CI

    </details>

11. Try to login again:

    <details>

    Screenshot 2022-03-16 at 20-50-13 Drone CI

    </details>

    You will: get a **half**-correct HTTP 500 error from Gitea.
   
    <details>

    Screenshot 2022-03-16 at 20-50-22 Gitea Git with a cup of tea

    </details>

    **You should be able to see this screen on the demo site at:** 
https://try.gitea.io/login/oauth/authorize?client_id=4d9b78ea-d8fc-45ed-a11a-ea684a690108&redirect_uri=https%3A%2F%2Fdrone1.kousu.ca%2Flogin&response_type=code&state=30b95ff183c471d4

13. If you can look in the database:

    You will: see the `oauth2_application` is still recorded.

    ```
    # sudo -u gitea psql -c 'select * from oauth2_application' | tee /dev/null
    id | uid |  name   |              client_id               |                        client_secret                         |           redirect_uris            | created_unix | updated_unix 
    ----+-----+---------+--------------------------------------+--------------------------------------------------------------+------------------------------------+--------------+--------------
      1 |   5 | DroneCI | 3cb8ec3b-5140-4875-bffa-716f9cf1cef7 | $2a$10$L6nyevnaNwKNB3cZc6rhUenodOlZ2DNFr4wIM.jPIP9BULRHRP8QW | ["http://3.98.145.180:8080/login"] |   1647449740 |   1647452962
    (1 row)
    ```

    even though the associated `uid` is gone:

    ```
    # sudo -u gitea psql -c 'select id,name from public.user where id=5;' | tee /dev/null
    id | name 
    ----+------
    (0 rows)
    ```
   

### Expected Behaviour

I would expect that deleting a user would delete all access rights that user had.

So, I would expect that deleting a user would delete all OAuth apps in their name **and immediately revoke** all tokens that were given out to it. The tokens not being revoked means a user can leave a ghost app around, at least

Right now you apparently can't get a new OAuth token issued (:+1:), but previously issued ones are still valid (:-1:), meaning someone who phished you into opening their malicious app can hide their tracks.

The HTTP 500 **should** instead be the standard HTTP 400 with "Client ID not registered":

Screenshot 2022-03-16 at 22-03-51 Gitea Git with a cup of tea

and clicking "Sync" on Drone (or any similar app) **should fail**.

### PR DESCRIPTION
**Pull request by 6543**
close  #19111

### CODE REVIEW

**Review by lunny — APPROVED** [review]

File: models/auth/oauth2.go [hunk 1]
```diff
@@ -576,3 +578,21 @@ func GetActiveOAuth2SourceByName(name string) (*Source, error) {
 
 	return authSource, nil
 }
+
+func DeleteOAuth2RelictsByUserID(ctx context.Context, userID int64) error {
+	deleteCond := builder.Select("id").From("oauth2_grant").Where(builder.Eq{"oauth2_grant.user_id": userID})
+
+	if _, err := db.GetEngine(ctx).In("grant_id", deleteCond).
+		Delete(&OAuth2AuthorizationCode{}); err != nil {
+		return err
```
Comment (Gusted) [inline models/auth/oauth2.go:587]: Also fmt.Errorf here?

File: models/auth/oauth2.go [hunk 2]
```diff
@@ -576,3 +578,21 @@ func GetActiveOAuth2SourceByName(name string) (*Source, error) {
...
+	}
+
+	if err := db.DeleteBeans(ctx,
+		&OAuth2Application{UID: userID},
+		&OAuth2Grant{UserID: userID},
+	); err != nil {
+		return fmt.Errorf("deleteBeans: %v", err)
```
Comment (delvh) [inline models/auth/oauth2.go]:
```suggestion
		return fmt.Errorf("DeleteBeans: %v", err)
```

File: models/auth/oauth2.go (hunk 1 above)
Comment (delvh) [reply models/auth/oauth2.go:587]: What would the responsible method be in that case?
`Delete`?

**Review by delvh — APPROVED** [review]

File: models/auth/oauth2.go (hunk 1 above)
Comment (6543) [reply models/auth/oauth2.go:587]: I do not expect an error here at all - so in the unlikely case just return it

**Pull request merged**